src/Interpreters/TreeRewriter.cpp

```diff
@@ -1030,16 +1030,18 @@ using RewriteShardNumVisitor = InDepthNodeVisitor<RewriteShardNum, true>;
 }
 
 TreeRewriterResult::TreeRewriterResult(
+    ContextPtr context_,
     const NamesAndTypesList & source_columns_,
     ConstStoragePtr storage_,
     const StorageSnapshotPtr & storage_snapshot_,
     bool add_special)
-    : storage(storage_)
+    : WithContext(context_)
+    , storage(storage_)
     , storage_snapshot(storage_snapshot_)
     , source_columns(source_columns_)
 {
     collectSourceColumns(add_special);
-    is_remote_storage = storage && storage->isRemote();
+    is_remote_storage = storage && storage->isRemote(getContext());
```

Comment:
I tried to find all the places where `is_remote_storage` is used:
* Tree optimizer, `optimizeMonotonousFunctionsInOrderBy`. There is a comment that we can't get the sorting key from the underlying table of distributed. But in case of parallel replicas,  I think, we can assume it's the same as in local replica.
* `TreeRewriterResult::collectUsedColumns`, to check if there is a virtual column `_shard_num`. I think we can assume there is no such column for parallel replicas.
* Inside InterpreterSelectQuery, analyze, `context->setDistributed`. But looks like it's only used for functions like `hostName()` to get different results from different shards/replicas. I think we can skip it for now for parallel replicas from MergeTree (and enable it later if needed).
* Inside ExpressionAnalyzer, to build subqueries for IN or GLOBAL IN. I suppose it's also ok for now to ignore GLOBAL IN for parallel replicas for now. As for IN, we avoid creating sets for Distributed tables. But it's actually ok to create a set earlier for parallel replicas cause there 100% should be a local replica, and a set for local replica should be built. And we won't build it a second time (it's better to check, but we shouldn't).

So, looks like all these cases are really only for Distributed. So I would vote to remove context from here).